en/03_Upgrading/07_Upgrading_Fluent.md

```diff
@@ -0,0 +1,306 @@
+
+![The "Localisation" action group, expanded to show all the actions available in it](../_images/fluent/batch_actions_enabled.png)
+
+[hint]
+Review available CMS UI components and enable those that you need.
+[/hint]
+
+## Fluent state
+
+Fluent state is the mechanism responsible for maintaining the global current locale state. Any time you swap between locales, the fluent state handles this change.
+
+In fluent 5 the way this state affects ORM actions changed.
+
+|              | **Fluent 4** | **Fluent 5** | **Fluent 6** |
+|--------------|--------------|--------------|--------------|
+| Fluent state | Local        | Global       | Global       |
+
+- **Local** - The current locale is set to the ORM query upon creation and
+    persists even if the global state change later in the code execution
+- **Global** - ORM query uses current locale from the global state at
+    the time of execution
+
+This affects any ORM calls inside a `withState()` callback which return a `DataList`. You should do a search in your codebase for calls to `withState()` and convert these to return the _result_ of the ORM operation instead of the `DataList`. For example:
+
+- return [`$list->column()`](api:SilverStripe/ORM/DataList::column()) or [`$list->getIDList()`](api:SilverStripe/ORM/DataList::getIDList())
+- return an aggregated result such as [`$list->min()`](api:SilverStripe/ORM/DataList::min()) or [`$list->max()`](api:SilverStripe/ORM/DataList::max())
+- return [`$list->getIterator()`](api:SilverStripe/ORM/DataList::getIterator())
+- return a specific record e.g. via [`$list->find()`](api:SilverStripe/ORM/DataList::find()) or [`$list->first()`](api:SilverStripe/ORM/DataList::first())
+- return an [`ArrayList`](api:SilverStripe/ORM/ArrayList) (e.g. `ArrayList::create($list->toArray())`)
+
+[warning]
+You should also check you don't have any situations where you set the state _without_ calling `withState()`, and then set it back again. These should be converted to use `withState()` instead.
+[/warning]
+
+### Code examples {#fluent-state-examples}
+
+#### Fluent v4 {#fluent-state-example-v4}
+
+```php
+$list = FluentState::singleton()->withState(function (FluentState $state): DataList {
+    $state->setLocale('my_locale_code');
+
+    return MyModel::get();
+});
+// The $list DataList will give results for the 'my_locale_code' locale, not whatever locale is in the global scope.
+$titles = $list->column('Title');
+```
+
+#### Fluent v5 and v6 {#fluent-state-example-v5-v6}
+
+```php
+$titles = FluentState::singleton()->withState(function (FluentState $state): array {
+    $state->setLocale('my_locale_code');
+
+    // This returns results for the 'my_locale_code' locale.
+    return MyModel::get()->column('Title');
+});
+// If you returned the DataList above, here it would give results for the locale in the global scope, not for 'my_locale_code'.
+```
+
```

Comment:
An example of how writing would work might also be handy.

#### Fluent v4 {#fluent-state-example-v4}

```php
// Set your current Locale state to locale_one before fetching the model
FluentState::singleton()->setLocale('locale_one');

// Fetching the model here will persist the Locale in the local state of the model
$model = MyModel::get()->byID(1);
$model->Title = 'Updated Title';

// Switching Locales here changes the global state, but not the local state of the model
FluentState::singleton()->setLocale('locale_two');

$model->write();
// The result is that "Updated Title" is saved to 'locale_one'
```

#### Fluent v5 and v6 {#fluent-state-example-v5-v6}

```php
// Set your current Locale state to locale_one before fetching the model
FluentState::singleton()->setLocale('locale_one');

// Fetching the model here sets the Locale to the local state of the model, but it is not persisted
$model = MyModel::get()->byID(1);
$model->Title = 'Updated Title';

// Switching Locales here changes both the global state, and the local state of the model
FluentState::singleton()->setLocale('locale_two');

$model->write();
// The result is that "Updated Title" is saved to 'locale_two'
```

Reply:
_would_ you temporarily set and revert the locale like that? It seems to me like that should be done inside `withState()` as well?

Reply:
Whether it's wrapped in `withState()` or not, the point I'm trying to get across is simply that the behaviour of the `write()` has changed.

`write()` behaviour was a significant change for some of our early projects, especially since the "patterns" that we now have were yet to be discovered/decided.

Reply:
Yup, that's a fair point. I'll add an example of writing once we've sorted out what has actually changed in https://github.com/silverstripe/developer-docs/pull/175#discussion_r1128327995

Reply:
Removed this section as per https://github.com/silverstripe/developer-docs/pull/175#discussion_r1128717103